iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert self["region"].dtype == pl.Utf8, "Column 'region' is not a string."
+        assert self["estimate"].dtype.is_numeric(), "Column 'estimate' is not numeric."
+        assert (
+            self["estimate"] >= 0.0
+        ).all(), "Cumulative uptake estimates cannot be negative."
+
+    def to_incident(self, rollout: dt.datetime) -> IncidentUptakeData:
+        self = (
+            self.with_columns(
+                season=pl.col("date").dt.year()
+                + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0),
+                estimate=pl.col("estimate")
+                .diff()
+                .fill_null(pl.col("estimate").first()),
+                elapsed=(pl.col("date") - rollout).cast(pl.Float64)
+                / (1000 * 60 * 60 * 24),
+            )
+            .with_columns(
+                interval=pl.col("elapsed").diff().fill_null(pl.col("elapsed").first()),
+                season=pl.col("season").cast(pl.Utf8),
+            )
+            .with_columns(estimate=pl.col("estimate") / pl.col("interval"))
+            .with_columns(previous=pl.col("estimate").shift(1))
+        )
+
+        if ((self["interval"] - self["interval"].mean()) / self["interval"].std())[
+            0
+        ] > 1:
+            self = self.slice(2, self.height - 1)
+        else:
+            self = self.slice(1, self.height - 1)
+
+        self = IncidentUptakeData(self)
+        self.validate()
+
+        return self
+
+
+class CumulativeUptakeProjection(UptakeData):
+    def validate(self):
+        assert set(self.columns).issuperset(
+            {"region", "date", "estimate"}
+        ), "At least one essential column is missing."
+        assert self["date"].dtype.is_temporal(), "Column 'date' is not temporal."
+        assert self["region"].dtype == pl.Utf8, "Column 'region' is not a string."
+        assert self["estimate"].dtype.is_numeric(), "Column 'estimate' is not numeric."
+
+
+def get_nis(
+    data_path, region_col, date_col, estimate_col, filters=None
+) -> CumulativeUptakeData:
+    out = (
+        (pl.read_csv(data_path))
+        .with_columns(estimate=pl.col(estimate_col).cast(pl.Float64, strict=False))
+        .drop_nulls(subset=["estimate"])
+    )
+
+    if filters is not None:
+        filter_expr = pl.lit(True)
+        for k, v in filters.items():
+            filter_expr &= pl.col(k).is_in(pl.lit(v))
+        out = out.filter(filter_expr)
+
+    out = (
+        out.with_columns(
+            region=pl.col(region_col),
+            date=pl.col(date_col)
+            .str.split(" ")
+            .list.get(0)
+            .str.strip_chars()
+            .str.to_date(format="%m/%d/%Y"),
+        )
+        .select(["region", "date", "estimate"])
+        .sort("region", "date")
+    )
+
+    out = CumulativeUptakeData(out)
+    out.validate()
+
+    return out
+
+
+class ProjectionSettings:
```

Comment:
Sounds like I have some learning to do on this front. But here's one question for now. Splitting data into train vs. test and sending commands to fit and evaluate are crucial steps. They may make sense as functions rather than methods that belong to an object, but why would we not want this function to still be part of the iup package?

Reply:
Fair question, and there's no "real" answer. In general CFA projects have had things like forecasting and evaluation *functions* in the package, but then the workflows that say "I want to forecast on all weeks between Oct 1 and Mar 1, then score all the results" in pipelines that are in places like `scripts/`.

So the idea is that the stuff in the package should be highly reusable, minimal actions like "forecast such-and-such dates using this this model and training data" or "give me the score if this was the prediction, this was the reality, and this is the scoring metric," and then have the scripts/pipelines be the things that are saying which datasets to look at, how many calls to be making, etc.